### PR TITLE
Remove dynamic allocations [8256]

### PR DESCRIPTION
**Pull request by MiguelCompany**
This PR removes the following dynamic allocations:

* `std::function` parameter on `readParameterListfromCDRMsg`
    * Implies moving ParameterList class to a private header
* Unnecessary vector copy on `WriterProxy::send`
* Unnecessary `CDRMessage` initialization when participant is not secure

### CODE REVIEW

File: src/cpp/rtps/writer/StatefulWriter.cpp [hunk 1]
```diff
@@ -1189,7 +1189,7 @@ bool StatefulWriter::matched_reader_add(
     {
         if (it->guid() == rdata.guid())
         {
-            logInfo(RTPS_WRITER, "Attempting to add existing reader, updating information." << endl);
+            logInfo(RTPS_WRITER, "Attempting to add existing reader, updating information." << std::endl);
```
Comment (IkerLuengo) [inline src/cpp/rtps/writer/StatefulWriter.cpp:1192]: No need to have an extra endline at the end of the log

File: src/cpp/rtps/writer/StatefulWriter.cpp [hunk 2]
```diff
@@ -1211,7 +1211,7 @@ bool StatefulWriter::matched_reader_add(
         else
         {
             logWarning(RTPS_WRITER, "Maximum number of reader proxies (" << max_readers << \
-                    ") reached for writer " << m_guid << endl);
+                    ") reached for writer " << m_guid << std::endl);
```
Comment (IkerLuengo) [inline src/cpp/rtps/writer/StatefulWriter.cpp:1214]: No need to have an extra endline at the end of the log

File: src/cpp/rtps/builtin/data/ParticipantProxyData.cpp [hunk 3]
```diff
@@ -33,10 +33,13 @@
 
 #include <rtps/builtin/data/ProxyHashTables.hpp>
 
+#include "../../../fastdds/core/policy/ParameterList.hpp"
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/rtps/builtin/data/ParticipantProxyData.cpp:36]: 👍

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/2033/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1784/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/2027/)

**Comment by MiguelCompany** [comment]
I cannot reproduce the warning on VS2015 😠 Will try again on Monday

**Comment by IkerLuengo** [comment]
It looks like std::function can throw std::bad_alloc on copy constructor. The old `ParameterList::readParameterListfromCDRMsg` took a std::function by value, and that may be the reason for the catch. Now that the method is templatized the compiler may not be able to detect a path to that exception.

I guess the try-catch can be removed, since we know the processor (defined right above the try) does not throw.

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/2060/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1816/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/2054/)

**Pull request merged**